File: parallel/ParallelWithoutCopy.dart
import 'dart:isolate';
import 'dart:async';

int sum(int LEN) {
  var res = 0;
  for (var i = 0; i < LEN; i++) {
    if (i % 2 == 0) res += i;
  }
  return res;
}

void psum(var msg) {
  int start = msg[0];
  int end = msg[1];
  SendPort replyTo = msg[2];
  int res = 0;
  for(var i = start;i < end;i++){
      if(i % 2 == 0)
          res += i;
  }
  //var res = a.where((n) => n % 2 == 0).reduce((a, b) => a + b);
  replyTo.send(res);
}

parallelSum(int LEN, int N,ReceivePort port) {
  var reply = port.sendPort;
  var offset = LEN ~/ N;
  for (var i = 0; i < N; i++) {
    Isolate.spawn(psum, [i * offset, i * offset + offset,reply]);
  }
}

main() async {
  final watch = new Stopwatch();
  int LEN = 5000000000;
  int N = 4;
  print("sum using loop");
  watch.start();
  print(sum(LEN));
  watch.stop();
  print("Elapsed time: ${watch.elapsedMilliseconds}ms");
  print("sum using parallel");
  watch..reset()..start();
  var res = 0;
  var count = 0;
  var port = new ReceivePort();
  parallelSum(LEN, N,port);
  port.listen((msg){
      res += msg;
      count++;
      if (count == N){
          print(res);
          port.close();
          watch.stop();
          print("Elapsed time: ${watch.elapsedMilliseconds}ms");
      }
  });
}

File: parallel/ParallelWithoutCopy.js
// ATTENTION:
// This file will be deprecated once node.js support transferale data.

var Worker = require('webworker-threads').Worker;
var now = require("performance-now");

var LEN = 30000000;
console.log('sum using loop');
var start = now();
console.log(sum(LEN));
var elapsed = now() - start;
console.log('Elapsed time: ' +  elapsed.toFixed(3) + 'ms');

console.log('sum using parallel');
var N = 4;  
var offset = LEN / N;
res = 0;
var count = 0;
start = now();
for(var i = 0;i < N;i++){
    var w =new Worker(psum);
    w.postMessage([i*offset,i*offset+offset]);
    w.onmessage = function (event) {
        res += event.data;
        count++;
        if(count == N){
            elapsed = now() - start;
            console.log(res);
            console.log('Elapsed time: ' +  elapsed.toFixed(3) + 'ms');
        }
    };
}

function sum(len){
    var res = 0;
    for(var i = 0;i < len;i++){
        if(i % 2 == 0) res += i;
    }
    return res;
}

function psum() {
    this.onmessage = function(e){
        var start = e.data[0],end = e.data[1];
        res = 0;
        for(var i = start;i < end;i++){
            if (i % 2 == 0) res += i;
        }
        postMessage(res);
        close();
    };
}
